### ISSUE TITLE
-Q option not working, 100% frame loss

### ISSUE THREAD

**Issue by jwbensley** [jwbensley/Etherate#5]
Via emai:

I try to use etherate and the -Q option. 
I run etherate on two x86 ends, but I have the result as :

One side :

root@ubuntu:~# ./etherate -I 2 -Q 1000 1000
Running in TX mode
Using device eth0 with address 00:00:60:cb:fd:ba, interface index 2
Entering promiscuous mode
Source MAC 00:00:5e:00:00:01
Destination MAC 00:00:5e:00:00:02
Sending gratuitous broadcasts...

Synchronising settings with RX host
Link quality tests enabled
Settings have been synchronised

Calculating delay between hosts...
Tx to Rx delay calculated as 0.000000051 seconds

Starting test on Fri Nov 4 12:19:09 2016

No. RTT Jitter
*
*
*
*
[...]
est frames transmitted: 30
Test frames received: 0
Non test or out-of-order frames received: 210
Number of timeouts: 31
Min/Max RTT during test: 999999.999999000s/0.000000000s
Min/Max jitter during test: 999999.999999000s/0.000000000s
Ending test on Fri Nov 4 12:19:39 2016

Leaving promiscuous mode
root@ubuntu:~#

Other side : 

oot@iou:~/SRC/Etherate-master# ./etherate -r
Running in RX mode
Using device eth0 with address 00:00:24:8b:26:3d, interface index 2
Entering promiscuous mode
Source MAC 00:00:5e:00:00:02
Destination MAC 00:00:5e:00:00:01
Sending gratuitous broadcasts...

Waiting for settings from TX host
Link quality tests enabled
Settings have been synchronised

Calculating delay between hosts...
Tx to Rx delay calculated as 0.000000051 seconds

Starting test on Fri Nov 4 12:22:06 2016
-> It never ends up.

Do you know why I am not able to have a correct test ?

**Comment by jwbensley** [comment]
The MTU test and one way delay/RTT test features are in their first draft, they have been coded but barely tested.

Looking through the code I found some hton() calls and ntoh() calls that were back to front.

I have corrected these in the latest commitment which seems to have fixed the problem however I noticed that with the quality/delay tests, if there is packet loss the output text on the CLI becomes out of order so there is more work to do here.

**Comment by jwbensley** [comment]
When calling `sendto()` on Tx and passing the frame transmit size as `FRAME_HEADERS->LENGTH+FRAME_HEADERS->TLV_SIZE` and if the receiver stored the incoming packet in the same memory location as the previous packet (it's random but seemed to be happening moderately often), the sub_tlv_type code from the previous frame was present in memory and aligned correctly so it was being read by Rx.

By setting to send a tlv+sub_tlv from Tx, Rx can re-use the same memory location and there is no misunderstanding of the received frame:

```
        // Let the receiver know all delay values have been received
        build_tlv(FRAME_HEADERS, htons(TYPE_TESTFRAME), htonl(VALUE_DELAY_END));

        build_sub_tlv(FRAME_HEADERS, htons(TYPE_TESTFRAME),
                      htonll(VALUE_DELAY_END));

        TX_RET_VAL = sendto(TEST_INTERFACE->SOCKET_FD,
                            FRAME_HEADERS->TX_BUFFER,
                            FRAME_HEADERS->LENGTH+FRAME_HEADERS->**SUB_TLV_SIZE**,
                            0,
                            (struct sockaddr*)&TEST_INTERFACE->SOCKET_ADDRESS,
                            sizeof(TEST_INTERFACE->SOCKET_ADDRESS));
```

**Comment by jwbensley** [comment]
Fixed in beta 0.11.
